plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -0,0 +1,375 @@
+import Data.Binary
+import Data.Hashable
+import qualified Data.HashMap.Strict as Map
+import Data.Maybe
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+import Data.Typeable
+import Development.IDE.Core.FileStore
+import Development.IDE.Core.OfInterest
+import Development.IDE.Core.Rules
+import Development.IDE.Core.Shake hiding (withIndefiniteProgress)
+import Development.IDE.Types.Diagnostics as D
+import Development.IDE.Types.Location
+import Development.Shake
+
+#ifdef GHC_LIB
+import Development.IDE.Core.RuleTypes (GhcSession(..))
+import Development.IDE.GHC.Util (hscEnv)
+import "ghc-lib" GHC hiding (DynFlags(..))
+import "ghc" GHC as RealGHC (DynFlags(..))
+import "ghc" HscTypes as RealGHC.HscTypes (hsc_dflags)
+import qualified "ghc" EnumSet as EnumSet
+import Language.Haskell.GhclibParserEx.GHC.Driver.Session as GhclibParserEx (readExtension)
+#else
+import GHC hiding (DynFlags(..))
+#endif
+
+import Ide.Logger
+import Ide.Types
+import Ide.Plugin
+import Ide.Plugin.Config
+import Ide.PluginUtils
+import Language.Haskell.HLint as Hlint
+import Language.Haskell.LSP.Core
+import Language.Haskell.LSP.Types
+import qualified Language.Haskell.LSP.Types      as LSP
+import qualified Language.Haskell.LSP.Types.Lens as LSP
+import System.FilePath (takeFileName)
+import System.IO (hPutStr, noNewlineTranslation, hSetNewlineMode, utf8, hSetEncoding, IOMode(WriteMode), withFile, hClose)
+import System.IO.Temp
+import Text.Regex.TDFA.Text()
+import GHC.Generics (Generic)
+
+-- ---------------------------------------------------------------------
+
+descriptor :: PluginId -> PluginDescriptor
+descriptor plId = (defaultPluginDescriptor plId)
+  { pluginRules = rules
+  , pluginCommands =
+      [ PluginCommand "applyOne" "Apply a single hint" applyOneCmd
+      , PluginCommand "applyAll" "Apply all hints to the file" applyAllCmd
+      ]
+    , pluginCodeActionProvider = Just codeActionProvider
+  }
+
+data GetHlintDiagnostics = GetHlintDiagnostics
+    deriving (Eq, Show, Typeable, Generic)
+instance Hashable GetHlintDiagnostics
+instance NFData   GetHlintDiagnostics
+instance Binary   GetHlintDiagnostics
+
+type instance RuleResult GetHlintDiagnostics = ()
+
+rules :: Rules ()
+rules = do
+  define $ \GetHlintDiagnostics file -> do
+    hlintOn' <- hlintOn <$> getClientConfigAction
+    logm $ "hlint:rules:hlintOn=" <> show hlintOn'
+    ideas <- if hlintOn' then getIdeas file else return (Right [])
+    return (diagnostics file ideas, Just ())
+
+  getHlintSettingsRule (HlintEnabled [])
+
+  action $ do
+    files <- getFilesOfInterest
+    void $ uses GetHlintDiagnostics $ Map.keys files
+
+  where
+
+      diagnostics :: NormalizedFilePath -> Either ParseError [Idea] -> [FileDiagnostic]
+      diagnostics file (Right ideas) =
+        [(file, ShowDiag, ideaToDiagnostic i) | i <- ideas, ideaSeverity i /= Ignore]
+      diagnostics file (Left parseErr) =
+        [(file, ShowDiag, parseErrorToDiagnostic parseErr)]
```

Comment:
Should we ignore these, leaving parse errors to be reported by normal Ghcide - otherwise we're likely to get two copies.

Reply:
If they match exactly the ones emitted by ghc, they should be removed, yeah

Reply:
Do they match exactly? My guess is I'm not following precisely the GHC formatting etc, since if I am, it would be by coincidence.

Reply:
I am observed duplicated errors for sure. The problem is if they dont match, hlint will stop to work and user will not know nothing about. 
The solution could be inspect the ghc warnings and no emit hlint ones if there is one from ghc in the same src span.
Not sure if it would be feasible, though.

Reply:
Ok, i take another path: when we have ghc parser errors, no hlint diagnostics is emitted. Only we'll have them when the user fix them all. But then we will emit hlint parsing errors, to inform user why hlint is not able to report real diagnostics.

I think it is simpler and the ux is not worse cause if users have to fix ghc parsing errors, hlint ones only can distract them.